### ISSUE TITLE
Internal structure of the metric of the metric store

### ISSUE THREAD

**Issue by ph** [elastic/logstash#4492]
When merging the JVM and the web api we found out that the current way of defining namespace and keys wasn't flexible enough.

Lets try to define all all the possibilities we can have.

# Top level
| metrics | type | phase | Status |
----------- | ------ | -------- | --------- |
stats.events.in | counter | 1 | Done in the PR, need to change namespace
stats.events.out | counter | 1 | Done in the PR, need to change namespace
stats.uptime_in_milliseconds | gauge value | 1
stats.logtash_version | gauge value | 1

| metrics | type | phase |
----------- | ------ | -------- |
stats.jvm.memory.heap.used_in_bytes | gauge value | 1
stats.jvm.memory.used_in_bytes | gauge | 1 
stats.jvm.memory.used_percent | gauge value | 1
stats.jvm.memory.committed_in_bytes | gauge | 1
stats.jvm.memory.max_in_bytes | gauge | 1

| metrics | type | phase |
----------- | ------ | -------- |
stats.os.cpu.XXX | X | X

# Detailed pipelines level, collected at the pipeline level
In the following case **XXX** is a **UUID** of user defined string in the logstash configuration.

| metrics | type | phase | status |
----------- | ------ | -------- | -------- |
stats.pipelines.pipeline_X.events.in | counter | 1  | Done in the PR, need to change namespace
stats.pipelines.pipeline_X.events.out | counter | 1 | Done in the PR, need to change namespace
stats.pipelines.pipeline_X.plugins.inputs.file_XXXXXX.events.out | counter |  1 | Done in the PR, need to change namespace
stats.pipelines.pipeline_X.plugins.filters.grok_XXXXXX.events.out | counter |  1 | Done in the PR, need to change namespace
stats.pipelines.pipeline_X.plugins.filters.grok_XXXXXX.events.in | counter |  1 | Done in the PR, need to change namespace
stats.pipelines.pipeline_X.plugins.outputs.elasticsearch_XXXX.events.in | counter |  1 | Done in the PR, need to change namespace
stats.pipelines.pipeline_X.plugins.outputs.elasticsearch_XXXX.events.out | counter |  need to be done at the plugin level or with a ack.

# Plugins Level metrics
Theses statistic are collected by the code inside the plugin and are defined by the plugin authors.

| metrics | type | phase |
----------- | ------ | -------- |
stats.pipelines.pipelines_X.plugins.outputs.elasticsearch_XXXX.error_count | counter | 2
stats.pipelines.pipelines_X.plugins.outputs.elasticsearch_XXXX.retry_count | counter | 2

# Query
All the keys are queryable on the metric store and return nested information when possible.
From a users point of vue they will be transformed into a hash.

| Query key | return |
| -------------- | ------- |
stats.pipelines | return all the pipelines information
stats.pipelines.pipeline_X  | return all the metrics from the pipeline X

For keeping thing simple in the first iteration the store doesn't support doing wildcards query,
but this could probably be done if needed in another iteration.

# API inside the pipelines or logstash core.
When you want to record a metric inside logstash, you have to provide a namespace and a key to save the corresponding metric.

In all the following scenarios it is recommended to use a variable when you are doing string interpolation. 

```ruby
metric.increment(:"stats.events", :in, 100)
metric.increment(:"stats.events", :in, 100)
metric.increment(:"stats.events", :out, 45)
metric.increment(:"stats.events.#{pipeline_id}.events", :in, 32)
metric.gauge(:"stats.jvm.memory", :free, 1200)
```

The previous example will create the following maps (divided by .)

```
stats -> events -> in
events["stats"]["events"]["in"]
```

If you want to make multiple calls on the same namespace  you can wrap the metric into a `NameSpacedMetric`
```ruby
namespaced = NameSpacedMetric.new(:stats.events, metric)
namespace.increment(:in, 100) 

#This code is the same as 
metric.increment(:"stats.events", :in, 100)
```
# API inside the plugins used by dev author.
Plugins authors need to be able to collect metric but we also need to scope where their metrics will be save. When a plugin is created we are giving the plugin instance a specific metric instance correctly scoped with the an initial namespace.

```ruby
namespaced_metric = NameSpacedMetric.new(:"stats.pipelines.pipelines_X.plugins.outputs.elasticsearch_XXXX.internal", metric)
namespaced_metric.increment(:error_count, 10) # will create stats.pipelines.pipelines_X.plugins.outputs.elasticsearch_XXXX.internal.error_count
```

# Metric Type
## Counter
In the current state of the code we only have the counter type which is a ever increasing counter.
We can use this to send data to elasticsearch and use the aggregation to retrieve the events per seconds in the system. No calculation is done on the logstash side. So the user of the api will only see the ever increasing values, they will need to do their own calculation.

## Gauge
More details in https://github.com/elastic/logstash/issues/4488
This values is mostly a snapshot in time of something, we can use this for memory usage, cpu usage and the uptime.
Again this value with the sampling will allow us to graph the evolutions over time if needed.

## Response Meter 
We will probably need something to collect response time, because a counter or a gauge make us lose way too much granularity. see https://github.com/elastic/logstash/issues/4497

**Comment by ph** [comment]
@purbon Can you review this before I do the changes, I think it address the issue of https://github.com/elastic/logstash/issues/4461

**Comment by purbon** [comment]
on it @ph

**Comment by purbon** [comment]
What does 

```stats.version => gauge value ``` mean? version of what? should this be collected by the collector?

**Comment by purbon** [comment]
for pipelines, see ```stats.pipelines.pipeline_X.plugins.input.file_XXXXXX.events.out => ever increasing counter``` I guess you missed one s for ```input```, right?

**Comment by purbon** [comment]
Why ```stats.pipelines.pipelines_X.plugins.outputs.elasticsearch_XXXX.internal.error_count ``` have internal? don't think is something we should differentiate, in the namespace, between metrics provided by plugins or by the logstash-core, what do you think?

**Comment by purbon** [comment]
Can you also add for the section ```API inside the pipelines or logstash core.``` how to create metrics namespaces? I find this important. 

There is also some room for improvements at my understanding, but would love to know more about form "the creator" before proposing them :+1: !

**Comment by purbon** [comment]
@ph would be nice to know if all this resources are actually provided by the current branch, in case they are not, would be nice to know which ones are "fase 1".

**Comment by ph** [comment]
> stats.version => gauge value mean? version of what? should this be collected by the collector?

The logstash version, renamed for `logstash_version` we will want to show that in marvel probably.

> I would move:
stats.jvm.memory.heap.used_in_bytes => gauge value
...
stats.jvm.memory.max_in_bytes => gauge
and
stats.os.cpu.XXX
to specific sections for jvm and os, what do you think?

I do not not understand what you mean here?

>for pipelines, see stats.pipelines.pipeline_X.plugins.input.file_XXXXXX.events.out => ever increasing counter I guess you missed one s for input, right?

Fixed the typo.

> Why stats.pipelines.pipelines_X.plugins.outputs.elasticsearch_XXXX.internal.error_count have internal? don't think is something we should differentiate, in the namespace, between metrics provided by plugins or by the logstash-core, what do you think?

Lets say I want to find all the metrics defined by the plugin author? I used that as a way to differentiate them.

**Comment by ph** [comment]
> There is also some room for improvements at my understanding, but would love to know more about form "the creator" before proposing them

Go ahead, ask the questions!

**Comment by ph** [comment]
I have updated the section "API inside the pipelines and/or logstash core with a bit more details how to use named space, Also added details for phase and the status, most `in/out` keys for plugins are done, I created this issue to make sure the keys will be already fixed :)

**Comment by acchen97** [comment]
Looks great @ph, we planning to collect internal queue stats for Phase 1 too?  Also, do you have the full list of all metrics we’ll actually be collecting here?

**Comment by ph** [comment]
@acchen97 That's a good catch I will add them!

**Comment by suyograo** [comment]
@ph why repeat the word pipeline here? Can't `stats.pipelines.pipeline_X.events.in` be renamed to `stats.pipeline_X.events.in` to compact it? Similarly `stats.pipelines.pipelines_X.plugins.outputs.elasticsearch_XXXX.error_count` becomes `stats.pipelines_X.outputs.elasticsearch_XXXX.error_count`

**Comment by ph** [comment]
@suyograo This was mostly to prevent collisions with named pipelines and namespace and make it easier to information retrieval.

This would make this kind of query possible to all the stats for all the pipelines.
```
QUERY stats.pipelines.*.events.in
```

**Comment by purbon** [comment]
@suyograo @ph yes to all said here. For now is the fastest approach to get this up and working, in later iteration we could mask this namespaces for the query face or the insert face by introducing a new layer, but internally I don't see us changing this structure soon as it clearly defines everything in their place.

**Comment by ph** [comment]
@acchen97 Well, the architecture changed a bit inside logstash since we originally though the api.

We arent using a fixed size queue inside logtash like in pre 2.2 version, we are now using a `SynchronousQueue` so technically only one item can be in the queue at any time, what we actually do is   take items in batch from the queue (internal buffering) and this batch size is fixed at startup time.
Not sure there is values in sampling that.

@andrewvc Correct me if I am wrong.

**Comment by ph** [comment]
merged in master
